yapapi/engine.py

```diff
@@ -547,9 +638,14 @@ async def worker_task(agreement: Agreement):
             work_context = WorkContext(activity, agreement, self.storage_manager, emitter=job.emit)
             work_context.emit(events.ActivityCreated)
 
+            self._activity_created_at[activity.id] = datetime.now()
+
             async with activity:
                 self.accept_debit_notes_for_agreement(job.id, agreement.id)
                 await run_worker(work_context)
+                # Providers may issue debit notes after activity ends.
+                # This will prevent terminating agreements when this happens.
+                self._activity_created_at.pop(activity.id, None)
```

Comment:
This will free some memory after activity ends:
```suggestion
                self._activity_created_at.pop(activity.id, None)
                self._num_debit_notes.pop(activity.id, None)
                self._num_payable_debit_notes.pop(activity.id, None)
```